library/core/src/iter/adapters/enumerate.rs

```diff
@@ -15,10 +16,131 @@ use crate::ops::{Add, AddAssign, Try};
 pub struct Enumerate<I> {
     iter: I,
     count: usize,
+    len: usize,
 }
-impl<I> Enumerate<I> {
+impl<I: Iterator> Enumerate<I> {
     pub(in crate::iter) fn new(iter: I) -> Enumerate<I> {
-        Enumerate { iter, count: 0 }
+        EnumerateImpl::new(iter)
+    }
+}
+
+// Enumerate specialization trait
+#[doc(hidden)]
+trait EnumerateImpl<I> {
+    type Item;
+    fn new(iter: I) -> Self;
+    fn next(&mut self) -> Option<Self::Item>;
+    unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess;
+    fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator;
+}
+
+impl<I> EnumerateImpl<I> for Enumerate<I>
+where
+    I: Iterator,
+{
+    type Item = (usize, I::Item);
+
+    default fn new(iter: I) -> Self {
+        Enumerate {
+            iter,
+            count: 0,
+            len: 0, // unused
+        }
+    }
+
+    #[inline]
+    default fn next(&mut self) -> Option<Self::Item> {
+        let a = self.iter.next()?;
+        let i = self.count;
+        // Possible undefined overflow.
+        AddAssign::add_assign(&mut self.count, 1);
+        Some((i, a))
+    }
+
+    #[inline]
+    default unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess,
+    {
+        // SAFETY: the caller must uphold the contract for
+        // `Iterator::__iterator_get_unchecked`.
+        let value = unsafe { try_get_unchecked(&mut self.iter, idx) };
+        (Add::add(self.count, idx), value)
+    }
+
+    #[inline]
+    default fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator,
+    {
+        let a = self.iter.next_back()?;
+        let len = self.iter.len();
+        // Can safely add, `ExactSizeIterator` promises that the number of
+        // elements fits into a `usize`.
+        Some((self.count + len, a))
+    }
+}
+
+// This is the same code as above but using `intrinsics::assume()` to hint at the compiler
+// that the returned index is smaller than the length of the underlying iterator.
+impl<I> EnumerateImpl<I> for Enumerate<I>
+where
+    I: TrustedLen + Iterator,
+{
+    fn new(iter: I) -> Self {
+        let len = iter.size_hint().0;
```

Comment:
What happens for iterators returning `(usize::MAX, None)` here?

Reply:
That's a very good question. I believe `TrustedLen` is actually too weak here and we always need `TrustedRandomAccess`.

With `TrustedLen` we could get `(usize::MAX, None)` and then don't know anything more about the length other than it being big. So the assumption would have to be `idx < len || len == usize::MAX` and that defeats the whole purpose of this PR.

`TrustedRandomAccess` OTOH gives us the guarantee that `len < usize::MAX`. That's not specifically written in the documentation of that trait (I'll add it now), but otherwise wouldn't work for its purpose (the specialized `Zip`).

Reply:
`ExactSizeIterator` + `TrustedLen` would work equally well. The former guarantees that the length fits into an `usize`, the latter guarantees that it's actually the length.

Reply:
Going for those two, that seems to cover slightly more iterators but all where I think this could become useful have both implemented anyway so...

Reply:
I suppose you could alternatively store an `Option<usize>` and use `size_hint().1`, and skip the assert on None. But that might make it slightly trickier for the optimizer.

Reply:
Yeah that seems problematic. I'm looking now into implemented `TrustedRandomAccess` in a few places that are `TrustedLen` but not `TrustedRandomAccess` yet. That should make the difference here smaller and also generally improve performance of those iterators so useful anyway.

I guess the main place where this can't be done (easily?) is for types like `BTreeMap` or `HashMap` that don't have an easy way of implemented `Iterator::__get_unchecked()`.